### PR TITLE
Add fields to @types/aws-lambda dynamo streams handler

### PR DESCRIPTION
**Pull request by jacsamell**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://docs.aws.amazon.com/lambda/latest/dg/with-ddb.html

### CODE REVIEW

**Review by simonbuchan — COMMENTED** [review]
Checking docs, it seems like the handler will know if these fields are going to be present based on whether tumbling windows are enabled. That would imply that it's actually better for this to be a separate event, handler etc. type, e.g. `DynamoDBStateStream{Event,...}`.

This is further attached to that whether you are required to return state, and that said state can be typed by the handler as it is presumably going to get back what it returned last.

Therefore, I would expect this to be leaving the existing types as is, and adding (with a strawman name):

```ts
export type DynamoDBStateStreamHandler<State> = Handler<DynamoDBStateStreamEvent<State>, DynamoDBStateStreamResult<State>>;
export type DynamoDBStateStreamCallback<State> = Callback<DynamoDBStateStreamResult<State>>;

export interface DynamoDBStateStreamEvent<State> {
    Records: DynamoDBRecord[];
    window: { start: string, end: string };
    state?: State; // or whatever AWS gives you for the first invoke? Empty object?
    isFinalInvokeForWindow: boolean;
    isWindowTerminatedEarly: boolean;
}

export interface DynamoDBStateStreamResult<State> {
    state: State;
}
```

Not sure that this works for you, let me know if the typing of state in particular makes sense: hard to tell from the docs!

**Comment by jacsamell** [comment]
> Checking docs, it seems like the handler will know if these fields are going to be present based on whether tumbling windows are enabled. That would imply that it's actually better for this to be a separate event, handler etc. type, e.g. `DynamoDBStateStream{Event,...}`.
> 
> This is further attached to that whether you are required to return state, and that said state can be typed by the handler as it is presumably going to get back what it returned last.
> 
> Therefore, I would expect this to be leaving the existing types as is, and adding (with a strawman name):
> 
> ```ts
> export type DynamoDBStateStreamHandler<State> = Handler<DynamoDBStateStreamEvent<State>, DynamoDBStateStreamResult<State>>;
> export type DynamoDBStateStreamCallback<State> = Callback<DynamoDBStateStreamResult<State>>;
> 
> export interface DynamoDBStateStreamEvent<State> {
>     Records: DynamoDBRecord[];
>     window: { start: string, end: string };
>     state?: State; // or whatever AWS gives you for the first invoke? Empty object?
>     isFinalInvokeForWindow: boolean;
>     isWindowTerminatedEarly: boolean;
> }
> 
> export interface DynamoDBStateStreamResult<State> {
>     state: State;
> }
> ```
> 
> Not sure that this works for you, let me know if the typing of state in particular makes sense: hard to tell from the docs!

Yes this makes perfect sense, will update today

File: types/aws-lambda/trigger/dynamodb-stream.d.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { Handler } from "../handler";
 
-export type DynamoDBStreamHandler = Handler<DynamoDBStreamEvent, void>;
+export type DynamoDBStreamHandler = Handler<DynamoDBStreamEvent, undefined | { state: any }>;
```
Comment (savnik) [inline types/aws-lambda/trigger/dynamodb-stream.d.ts:3]:
```suggestion
export type DynamoDBStreamHandler = Handler<DynamoDBStreamEvent, undefined | { state: any }>;
```

type any is disregarded, for unknown maps use: `Map<String, String>` as suggested in the AWS docuemntation

**Review by savnik — CHANGES_REQUESTED** [review]

**Review by darbio — CHANGES_REQUESTED** [review]
As per comments. Please update PR to reflect the requested changes.

**Comment by doroodgar** [comment]
hi

در تاریخ سه‌شنبه ۲۲ فوریهٔ ۲۰۲۲،‏ ۱۰:۲۱ TypeScript Bot <
***@***.***> نوشت:

**Pull request closed without merging**